dbms/src/Interpreters/JoinV2/HashJoinProbeBuildScanner.cpp

```diff
@@ -0,0 +1,411 @@
+        break;
+        APPLY_FOR_HASH_JOIN_VARIANTS(M)
+#undef M
+
+    default:
+        throw Exception(
+            fmt::format("Unknown JOIN keys variant {} during scanning build side", magic_enum::enum_name(join->method)),
+            ErrorCodes::UNKNOWN_SET_DATA_VARIANT);
+    }
+
+#undef CALL
+#undef CALL2
+#undef SET_FUNC_PTR
+}
+
+Block JoinProbeBuildScanner::scan(JoinProbeWorkerData & wd)
+{
+    return (this->*scan_func_ptr)(wd);
+}
+
+template <typename KeyGetter, ASTTableJoin::Kind kind, bool need_row_data, bool need_other_block_data>
+Block JoinProbeBuildScanner::scanImpl(JoinProbeWorkerData & wd)
+{
+    static_assert(need_row_data || need_other_block_data);
+
+    if (wd.is_scan_end)
+        return {};
+
+    using KeyGetterType = typename KeyGetter::Type;
+    using HashValueType = typename KeyGetter::HashValueType;
+    const auto & multi_row_containers = join->multi_row_containers;
+    const size_t max_block_size = join->settings.max_block_size;
+    const size_t left_columns = join->left_sample_block_pruned.columns();
+    auto & non_joined_blocks = join->non_joined_blocks;
+
+    auto & key_getter = *static_cast<KeyGetterType *>(join_key_getter.get());
+    constexpr size_t key_offset
+        = sizeof(RowPtr) + (KeyGetterType::joinKeyCompareHashFirst() ? sizeof(HashValueType) : 0);
+
+    Block * full_block = non_joined_blocks.getNextFullBlock();
+    if (full_block != nullptr)
+        return *full_block;
+
+    size_t scan_size = 0;
+    RowContainer * container = wd.current_container;
+    size_t index = wd.current_container_index;
+    wd.selective_offsets.clear();
+    wd.selective_offsets.reserve(max_block_size);
+    constexpr size_t insert_batch_max_size = 256;
+    wd.insert_batch.clear();
+    wd.insert_batch.reserve(insert_batch_max_size);
+    join->initOutputBlock(wd.scan_result_block);
+
+    Block * non_joined_non_full_block = nullptr;
+    size_t output_columns = wd.scan_result_block.columns();
+    while (true)
+    {
+        non_joined_non_full_block = non_joined_blocks.getNextNonFullBlock();
+        if (non_joined_non_full_block == nullptr)
+            break;
+        RUNTIME_CHECK(non_joined_non_full_block->columns() == output_columns);
+        size_t rows = non_joined_non_full_block->rows();
+        if (rows >= max_block_size / 2)
+            return *non_joined_non_full_block;
+
+        if (wd.scan_result_block.rows() + rows > max_block_size)
+        {
+            Block res_block;
+            res_block.swap(wd.scan_result_block);
+
+            join->initOutputBlock(wd.scan_result_block);
+            for (size_t i = 0; i < output_columns; ++i)
+            {
+                auto & src_column = non_joined_non_full_block->getByPosition(i);
+                auto & des_column = wd.scan_result_block.getByPosition(i);
+                des_column.column->assumeMutable()->insertRangeFrom(*src_column.column, 0, rows);
+            }
+            return res_block;
```

Comment:
why return directly, no need to insert default for left table for RightOuterJoin?

Reply:
The schema of `non_joined_non_full_block` is `HashJoin::output_block_after_finalize`. The default values are already inserted into the columns in the left table of `non_joined_non_full_block` during build phase.